File: app/src/main/java/com/abishov/hexocat/common/network/NetworkModule.java
package com.abishov.hexocat.common.network;

import android.content.Context;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.TypeAdapterFactory;
import dagger.Module;
import dagger.Provides;
import java.util.concurrent.TimeUnit;
import javax.inject.Singleton;
import okhttp3.Cache;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.CallAdapter;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

@Module
@Singleton
public abstract class NetworkModule { // NOPMD

  private static final String OK_HTTP = "OkHttp";

  @Provides
  @Singleton
  static TypeAdapterFactory gsonAdapterFactory() {
    return HexocatAdapterFactory.create();
  }

  @Provides
  @Singleton
  static CallAdapter.Factory rxJavaAdapterFactory() {
    return RxJava2CallAdapterFactory.create();
  }

  @Provides
  @Singleton
  static Gson gson(TypeAdapterFactory adapterFactory) {
    return new GsonBuilder()
        .registerTypeAdapterFactory(adapterFactory)
        .create();
  }

  @Provides
  @Singleton
  static Cache okHttpCache(Context context) {
    return new Cache(context.getCacheDir(), 10 * 1024 * 1024); // 10 MiBs
  }

  @Provides
  @Singleton
  static Interceptor okHttpLogging(Cache cache) {
    return new HttpLoggingInterceptor(message -> {
      Timber.tag(OK_HTTP).d(message);
      Timber.tag(OK_HTTP).v("Cache: requests=[%s], network=[%s], hits=[%s]",
          cache.requestCount(), cache.networkCount(), cache.hitCount());
    }).setLevel(HttpLoggingInterceptor.Level.BASIC);
  }

  @Provides
  @Singleton
  static OkHttpClient okHttpClient(Interceptor logger, Cache cache) {
    return new OkHttpClient.Builder()
        .addInterceptor(logger)
        .cache(cache)
        .connectTimeout(32, TimeUnit.SECONDS)
        .writeTimeout(32, TimeUnit.SECONDS)
        .readTimeout(32, TimeUnit.SECONDS)
        .build();
  }

  @Provides
  @Singleton
  static Converter.Factory gsonFactory(Gson gson) {
    return GsonConverterFactory.create(gson);
  }

  @Provides
  @Singleton
  static Retrofit retrofit(HttpUrl baseUrl, Converter.Factory factory,
      CallAdapter.Factory rxJavaAdapterFactory,
      OkHttpClient okHttpClient) {
    return new Retrofit.Builder()
        .baseUrl(baseUrl)
        .addConverterFactory(factory)
        .addCallAdapterFactory(rxJavaAdapterFactory)
        .client(okHttpClient)
        .build();
  }
}

File: app/src/main/java/com/abishov/hexocat/common/network/NetworkModule.kt
package com.abishov.hexocat.common.network

import android.content.Context
import android.net.Uri
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.api.CustomTypeAdapter
import com.github.type.CustomType
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoSet
import okhttp3.Cache
import okhttp3.HttpUrl
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit
import javax.inject.Singleton
import javax.net.ssl.SSLSocketFactory
import javax.net.ssl.X509TrustManager

@Module
class NetworkModule {

  @Provides
  @Singleton
  fun okHttpCache(context: Context): Cache {
    return Cache(context.cacheDir, (10 * 1024 * 1024).toLong()) // 10 MiBs
  }

  @Provides
  @Singleton
  fun httpLogger(cache: Cache): HttpLoggingInterceptor.Logger = HttpLogger(cache)

  @IntoSet
  @Provides
  @Singleton
  fun authenticationInterceptor(): Interceptor = AuthInterceptor()

  @IntoSet
  @Provides
  @Singleton
  fun okHttpLogging(logger: HttpLoggingInterceptor.Logger): Interceptor {
    val interceptor = HttpLoggingInterceptor(logger)

    return interceptor.apply {
      level = HttpLoggingInterceptor.Level.BODY
    }
  }

  @Provides
  @Singleton
  fun okHttpClient(
    cache: Cache,
    interceptors: Set<@JvmSuppressWildcards Interceptor>,
    sslSocketFactory: SSLSocketFactory?,
    trustManager: X509TrustManager?
  ): OkHttpClient {
    val builder = OkHttpClient.Builder()
      .cache(cache)
      .connectTimeout(32, TimeUnit.SECONDS)
      .writeTimeout(32, TimeUnit.SECONDS)
      .readTimeout(32, TimeUnit.SECONDS)

    interceptors.forEach { builder.addInterceptor(it) }

    if (sslSocketFactory != null && trustManager != null) {
      builder.sslSocketFactory(sslSocketFactory, trustManager)
    }

    return builder.build()
  }

  @Provides
  @Singleton
  fun uriTypeAdapter(): CustomTypeAdapter<Uri> = UriTypeAdapter()

  @Provides
  @Singleton
  fun apolloClient(
    baseUrl: HttpUrl,
    client: OkHttpClient,
    uriTypeAdapter: CustomTypeAdapter<Uri>
  ): ApolloClient {
    return ApolloClient.builder()
      .addCustomTypeAdapter(CustomType.URI, uriTypeAdapter)
      .okHttpClient(client)
      .serverUrl(baseUrl)
      .build()
  }
}
